app/controllers/mixins/ems_common/angular.rb

```diff
@@ -775,10 +788,14 @@ def build_credentials(ems, mode)
           console_password = params[:console_password] ? params[:console_password] : ems.authentication_password(:console)
           creds[:console] = {:userid => params[:console_userid], :password => console_password, :save => (mode != :validate)} # FIXME: skateman was here
         end
-        if ems.kind_of?(ManageIQ::Providers::Amazon::CloudManager) &&
-           ems.supports_authentication?(:smartstate_docker) && params[:smartstate_docker_userid]
-          smartstate_docker_password = params[:smartstate_docker_password] ? params[:smartstate_docker_password] : ems.authentication_password(:smartstate_docker)
-          creds[:smartstate_docker] = {:userid => params[:smartstate_docker_userid], :password => smartstate_docker_password, :save => true}
+        if ems.kind_of?(ManageIQ::Providers::Amazon::CloudManager)
```

Comment:
Same as above.

Reply:
What `supports?` should be here? As you see, here we have two sub-conditionals.
Should the first one be `supports_smartstate_docker?` or, considering, that we check for Amazon before, should it be `supports_amazon_smartstate_docker?` In the first case we doesn't get rid of Amazon condition, in the second one - we just moving mess from one place to another, as it will be the same Amazon-specific solution, plugability makes no sense here.

Reply:
@AlexanderZagaynov : the design of the `supports_?` methods is fully in the hands of the backend.

Regarding "moving the mess"  -- this way the mess is in the UI and the mess contains a business logic that the UI should have no knowledge of.

Imagine that moving forward this `.rb` file should be on the client as `.js`. Do we want this kind of logic to be running in the browser?

I do not want that. We need to know a set of capabilities and pass a set of capabilities to the UI so that correct UI elements for the capabilities are displayed. We do not want to decide based on the Provider name (or somewhere I even saw substring of the provider name).

With the "supports_" based design you add a capability on the backend to a provider that previously lacked the capability and the UI works.

With the testing the provider name approach, you need to update all the places, have the knowledge spread between the backend, API, UI. Javascript, Ruby. A recipe for a regression.

Reply:
@martinpovolny let's sort flies from the meatballs. What you want requires a lot of work without predictable result. If you want to have this code refactored - let's do special BZ and ask @Loicavenel does it worth to do. Otherwise you just proposing to create more mess - part of the code in the old way, part - in the new, and half a year later someone else will do something third?
What I see now, that you turning simple task to add field into an epic story for 2-3 months or more.

Reply:
I do not want you to refactor this form. I want you to use a mechanism that was agreed a long time ago -- the `supports_?` -- the capabilities.

You do not need to fix this. Someone from the UI will have to do that. You are just required not to make it worse.

Reply:
done